Some unit tests! Testing the downloader.

rjw57 opened the pull request:
OK, so there's a lot here. Essentially the PR has three parts:

1) Extend the downloader with the ability to save the original TIFF files. This isn't crucial functionality but if you want to re-generate the mock-data used in the test suite you'll need a similar tool. If `--split-chunks` is passed on the command line, the individual .tif files are saved rather than being concatenated into a single dataset. This is a bit of a hack but I found it useful.

2) Use the [responses](https://github.com/dropbox/responses) library from dropbox to mock the HTTP calls. We ship a (very) small set of 8x8 pixel TIFF DEM files down-sampled from the real data. This is a binary blob but it's only 13KiB. It can be re-generated using a script in the mock data directory. With the mock data we can make something like the DEM download server appear in the test suite without having to a) download 7GiB of data and b) actually run a HTTP server. This mock data can probably be used in the fullness of time to test the actual elevation API.

3) Use the mock server to exercise the downloader a little in the test suite. This required teaching the downloader about varying data sizes but that change is pretty small.

I've also added the plumbing to post code-coverage information to coveralls.io if it's set up for the repo.

Overall this is a lot of code churn to a tool which is used precisely once when setting up the server but it's all infrastructure which can now be used to actually test the _API_ rather than just the download tool :).

rjw57 committed:
download: add option to download separate chunks

This is a little ugly in that it sort of hacks two different semantics
into one option: <dataset-location>. If we're to transition to a
different dataset storage format we need the tools to speak both worlds
for the moment.

rjw57 committed:
download: set default host via a constant

Use the same interpolation trick as used for default download location.

rjw57 committed:
download: parse chunk spec in main()

To aid testing, make download() be more of an "API" than a "command line
parser".

rjw57 committed:
download: optionally keep original chunk filenames

rjw57 committed:
tests: add simple test of downloader

This adds a test which tries to download two DEM chunks. USe the
responses mock framework to avoid trying to download ~80MiB of data on
each test run(!)

rjw57 committed:
tests: add mock DEM data

This data is a zip file containing GeoTIFF DEM chunks downsampled from
our "real" data into 8x8 pixel TIFFs. The resulting zip file is only
13KiB. The generate-dem-chunks.sh script may be used to re-create the
zip file from a local cache of the full DEM data.

rjw57 committed:
tests: add requirements for test suite and update travis

Update travis config to use the explicit requirements listed in
tests/requirements.txt.

rjw57 committed:
download: protect use of file object in with statement

Do not rely on CPython's behaviour w.r.t. file objects.

rjw57 committed:
download: allow expected size to be customised

In the test suite, we mock in some data to be downloaded with a
different size. Allow customisation of the expected size.

rjw57 committed:
tests: expand downloader tests

Expand the simple downloader test to try all chunks and to try the
support for downloading chunk files directly.

rjw57 committed:
tests: add README for mock data

rjw57 committed:
tests: add copyright banner to new files

I forgot to add the copyright banners to all the tests. Add them in.

rjw57 committed:
tests:travis: add integration with coveralls.io

Coveralls.io can be used to track code-coverage stats over time. We're
currently running at around 70% coverage which isn't bad but we can do
better.

rjw57 committed:
tests:travis: add integration with coveralls.io

Coveralls.io can be used to track code-coverage stats over time. We're
currently running at around 70% coverage which isn't bad but we can do
better.

rjw57 committed:
tests:download: collect common code together

There was all-together too much copy-and-paste code in testdownload.py.
Collect common code together.

adamgreig commented:
I think this all looks fine, but haven't had enough time to go through it properly yet, sorry. Unless I hear otherwise ( @danielrichman ) I'll try and go through and merge tomorrow.

rjw57 committed:
test:download: add test for main()

Add a simple test which mocks sys.argv allowing for full end-to-end
testing of the downloader.

rjw57 committed:
download: fix breakage of --split-chunks

The use of ``with target_f`` in download.py was invalid if target_f is
None. Add a somewhat ughly workaround and a test.

(This shows the advantage of nearly-full-stack testing...)

rjw57 commented:
Just added a few extra commits which test the actuall command-line parsing and a fix for a bug discovered by said.

rjw57 committed:
downloader: verify resolution string

Add some sanity checks to make sure that the resolution string is
sensibly formed and some tests which exercise this.
